Cannot find module '../settings'

Following instructions in README, did this:

> npm install -g agenda-rest
> agenda-rest --dbhost my.ip.here --dbname agenda

and get the following:

module.js:538
    throw err;
    ^

Error: Cannot find module '../settings'
    at Function.Module._resolveFilename (module.js:536:15)
    at Function.Module._load (module.js:466:25)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/usr/lib/node_modules/agenda-rest/dist/index.js:1:1547)
    at o (/usr/lib/node_modules/agenda-rest/dist/index.js:1:422)
    at Object.<anonymous> (/usr/lib/node_modules/agenda-rest/dist/index.js:1:4530)
    at o (/usr/lib/node_modules/agenda-rest/dist/index.js:1:422)
    at Object.defineProperty.value (/usr/lib/node_modules/agenda-rest/dist/index.js:1:843)
    at /usr/lib/node_modules/agenda-rest/dist/index.js:1:852

Am I missing something that's not included in the 'npm install' ?

OS: Ubuntu 16.04 LTS
Node: 8.9.3
NPM: 5.8.0

--- a/.babelrc
+++ b/.babelrc
@@ -5,7 +5,7 @@
     "presets": [
       ["@babel/preset-env", {
         "targets": {
-          "node": "current"
+          "node": "7.10"
         }
       }]
     ]

--- a/package.json
+++ b/package.json
@@ -62,7 +62,7 @@
     "koa-bodyparser": "^4.2.0",
     "koa-logger": "^3.1.0",
     "koa-router": "^7.0.1",
-    "pythonic": "0.0.6",
+    "pythonic": "^0.2.1",
     "regenerator-runtime": "^0.11.1",
     "request": "^2.79.0",
     "request-promise": "^4.1.1",

--- a/package.json
+++ b/package.json
@@ -18,6 +18,7 @@
   },
   "files": [
     "cli.js",
+    "settings.js",
     "dist"
   ],
   "keywords": [

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "agenda-rest",
-  "version": "0.7.3",
+  "version": "0.7.4",
   "description": "Scheduling as a Service",
   "main": "./dist/index.js",
   "jsnext:main": "./src/index.js",

--- a/package.json
+++ b/package.json
@@ -74,6 +74,7 @@
   "xo": {
     "space": 2,
     "rules": {
+      "promise/param-names": 0,
       "max-params": [
         "error",
         5

--- a/src/index.js
+++ b/src/index.js
@@ -41,11 +41,11 @@ router.get('/api/job', async (ctx, next) => {
   await next();
 });
 
-router.post('/api/job', getJobMiddleware(jobAssertions.notExists, jobOperations.define));
+router.post('/api/job', getJobMiddleware(jobAssertions.notExists, jobOperations.create));
 
 router.del('/api/job/:jobName', getJobMiddleware(jobAssertions.alreadyExists, jobOperations.delete));
 
-router.put('/api/job/:jobName', getJobMiddleware(jobAssertions.alreadyExists, jobOperations.define));
+router.put('/api/job/:jobName', getJobMiddleware(jobAssertions.alreadyExists, jobOperations.update));
 
 router.post('/api/job/once', getJobMiddleware(jobAssertions.alreadyExists, jobOperations.once));
 

--- a/src/index.js
+++ b/src/index.js
@@ -66,5 +66,5 @@ const graceful = () => {
 process.on('SIGTERM', graceful);
 process.on('SIGINT', graceful);
 
-export {app, router, jobsReady};
+export {app, router, agenda, jobsReady};
 export default app;

--- a/src/job.js
+++ b/src/job.js
@@ -27,7 +27,8 @@ const jobAssertions = {
   doNotAssert: () => true
 };
 
-const defineJob = async ({name, url, method, callback} = {}, jobs, agenda) => {
+const defineJob = async (job, jobs, agenda) => {
+  const {name, url, method, callback} = job;
   agenda.define(name, (job, done) => {
     const {attrs: {data}} = job;
     let uri = url;
@@ -75,7 +76,7 @@ const defineJob = async ({name, url, method, callback} = {}, jobs, agenda) => {
   });
 
   await jobs.count({name})
-    .then(count => count < 1 ? jobs.insert({name, url, method, callback}) : jobs.update({name}, {$set: {url, method, callback}}));
+    .then(count => count < 1 ? jobs.insert(job) : jobs.update({name}, {$set: job}));
 
   return 'job defined';
 };
@@ -84,7 +85,7 @@ const deleteJob = async (job, jobs, agenda) => {
   const cancel = promisify(agenda.cancel).bind(agenda);
   const numRemoved = await cancel(job);
   const obj = await jobs.remove(job);
-  return `removed ${numRemoved} job definitions and ${obj.result.n} job instances.`;
+  return `removed ${obj.result.n} job definitions and ${numRemoved} job instances.`;
 };
 
 const cancelJob = async (job, jobs, agenda) => {
@@ -134,7 +135,8 @@ const getScheduleJobFunction = scheduleType => async (job, jobs, agenda) => {
 const getJobOperation = (checkFunction, jobFunction) => ({check: checkFunction, fn: jobFunction});
 
 const jobOperations = {
-  define: getJobOperation(getCheckJobFormatFunction('url'), defineJob),
+  create: getJobOperation(getCheckJobFormatFunction('url'), defineJob),
+  update: getJobOperation(getCheckJobFormatFunction(), defineJob),
   delete: getJobOperation(getCheckJobFormatFunction(), deleteJob),
   cancel: getJobOperation(doNotCheck, cancelJob),
   now: getJobOperation(

--- a/src/job.js
+++ b/src/job.js
@@ -103,12 +103,12 @@ const getDefaultJobForSchedule = () => ({
 
 const scheduleTypes = {
   now: {
-    method: agenda => promisify(agenda.now).bind(agenda),
+    fn: agenda => promisify(agenda.now).bind(agenda),
     message: 'for now',
     getParams: job => [job.name, job.data]
   },
   once: {
-    method: agenda => promisify(agenda.schedule).bind(agenda),
+    fn: agenda => promisify(agenda.schedule).bind(agenda),
     message: 'for once',
     getParams: job => {
       // Check if interval is timestamp
@@ -121,14 +121,14 @@ const scheduleTypes = {
     }
   },
   every: {
-    method: agenda => promisify(agenda.every).bind(agenda),
+    fn: agenda => promisify(agenda.every).bind(agenda),
     message: 'for repetition',
     getParams: job => [job.interval, job.name, job.data]
   }
 };
 
 const getScheduleJobFunction = scheduleType => async (job, jobs, agenda) => {
-  await scheduleType.method(agenda)(...scheduleType.getParams(job));
+  await scheduleType.fn(agenda)(...scheduleType.getParams(job));
   return `job scheduled ${scheduleType.message}`;
 };
 

--- a/src/job.js
+++ b/src/job.js
@@ -31,13 +31,20 @@ const defineJob = async ({name, url, method, callback} = {}, jobs, agenda) => {
   agenda.define(name, (job, done) => {
     const {attrs: {data}} = job;
     let uri = url;
+    // http://example.com/foo/:param1/:param2
+    // =>
+    // http://example.com/foo/value1/value2
     for (const [key, value] of keyValues(data.params)) {
       uri = uri.replace(`:${key}`, value);
     }
+    // http://example.com/foo
+    // =>
+    // http://example.com/foo?query1=value1&query2=value2
     const query = querystring.stringify(data.query);
     if (query !== '') {
       uri += `?${query}`;
     }
+    // Error if no response in timeout
     Promise.race([
       new Promise((resolve, reject) => setTimeout(() => reject(new Error('TimeOutError')), settings.timeout)),
       rp({
@@ -58,7 +65,7 @@ const defineJob = async ({name, url, method, callback} = {}, jobs, agenda) => {
             method: callback.method || 'POST',
             uri: callback.url,
             headers: callback.headers || {},
-            body: {data: data.body, response: result},
+            body: {data, response: result},
             json: true
           });
         }

--- a/src/util.js
+++ b/src/util.js
@@ -25,6 +25,18 @@ const bootstrapKoaApp = () => {
 
 const isValidDate = date => Object.prototype.toString.call(date) === '[object Date]' && !isNaN(date.getTime());
 
+const repeatPerKey = (keys = {}) => count => (key, fn) => () => {
+  if (!(key in keys)) {
+    keys[key] = 0;
+  }
+  if (keys[key] < count) {
+    fn();
+    keys[key]++;
+  }
+};
+
+const oncePerKey = repeatPerKey()(1);
+
 class AsyncCounter {
   constructor(countTimes) {
     let currentCount = 0;
@@ -45,4 +57,4 @@ class AsyncCounter {
   }
 }
 
-export {bootstrapKoaApp, isValidDate, AsyncCounter};
+export {bootstrapKoaApp, isValidDate, oncePerKey, AsyncCounter};

--- a/src/util.js
+++ b/src/util.js
@@ -26,15 +26,19 @@ const bootstrapKoaApp = () => {
 const isValidDate = date => Object.prototype.toString.call(date) === '[object Date]' && !isNaN(date.getTime());
 
 class AsyncCounter {
-  constructor(countTo) {
+  constructor(countTimes) {
+    let currentCount = 0;
+    this.countTimes = countTimes;
     this.ready = new Promise(resolveReady => {
       this.finished = new Promise(resolveFinished => {
-        this.count = () => {
-          countTo--;
-          if (countTo === 0) {
+        const count = () => {
+          currentCount++;
+          if (currentCount === countTimes) {
             resolveFinished();
           }
+          return currentCount;
         };
+        this.count = () => this.ready.then(() => count());
         resolveReady();
       });
     });

--- a/src/util.js
+++ b/src/util.js
@@ -25,4 +25,20 @@ const bootstrapKoaApp = () => {
 
 const isValidDate = date => Object.prototype.toString.call(date) === '[object Date]' && !isNaN(date.getTime());
 
-export {bootstrapKoaApp, isValidDate};
+class AsyncCounter {
+  constructor(countTo) {
+    this.ready = new Promise(resolveReady => {
+      this.finished = new Promise(resolveFinished => {
+        this.count = () => {
+          countTo--;
+          if (countTo === 0) {
+            resolveFinished();
+          }
+        };
+        resolveReady();
+      });
+    });
+  }
+}
+
+export {bootstrapKoaApp, isValidDate, AsyncCounter};

--- a/test.js
+++ b/test.js
@@ -1,8 +1,7 @@
 const {promisify} = require('util');
 const test = require('ava');
 const request = require('supertest');
-const {range} = require('pythonic');
-const {bootstrapKoaApp} = require('./src/util');
+const {bootstrapKoaApp, AsyncCounter} = require('./src/util');
 
 const agendaAppUrl = 'http://localhost:4041';
 const testAppUrl = 'http://localhost:4042';
@@ -55,16 +54,16 @@ test.serial('PUT /api/job succeeds when the job exists', async t => {
   t.is(res.status, 200);
 });
 
-const testAppCalled = async (defineEndpoint, thisManyTimes = 1) => {
-  const calls = range(thisManyTimes).map(() => new Promise(resolve => defineEndpoint(resolve)));
-  return Promise.all(calls);
-};
+let counter;
 
-const defineEndpointFoo = done => testAppRouter.post('/foo', async (ctx, next) => {
+testAppRouter.post('/foo', async (ctx, next) => {
   console.log('foo invoked!');
+  console.log(new Date());
   ctx.body = 'foo success';
   ctx.status = 200;
-  done();
+  if (counter) {
+    counter.ready.then(() => counter.count());
+  }
   await next();
 });
 /* TODO
@@ -88,25 +87,25 @@ const defineEndpointFooCallback = done => testAppRouter.post('/foo/cb', async (c
 */
 
 test.serial('POST /api/job/now with existing foo definition invokes the foo endpoint', async t => {
-  const fooInvoked = testAppCalled(defineEndpointFoo);
+  counter = new AsyncCounter(1);
   const res = await agendaAppRequest
     .post('/api/job/now')
     .send({name: 'foo'});
 
   t.is(res.text, 'job scheduled for now');
 
-  return fooInvoked;
+  return counter.finished;
 });
 
 test.serial('POST /api/job/every with existing foo definition invokes the foo endpoint', async t => {
-  const fooInvoked = testAppCalled(defineEndpointFoo, 3);
+  counter = new AsyncCounter(3);
   const res = await agendaAppRequest
     .post('/api/job/every')
-    .send({name: 'foo', interval: '10 miliseconds'});
+    .send({name: 'foo', interval: '2 seconds'});
 
   t.is(res.text, 'job scheduled for repetition');
 
-  return fooInvoked;
+  await counter.finished;
 });
 
 test.serial('DELETE /api/job succeeds when a job is defined', async t => {

--- a/test.js
+++ b/test.js
@@ -1,6 +1,7 @@
 const {promisify} = require('util');
 const test = require('ava');
 const request = require('supertest');
+const {range} = require('pythonic');
 const {bootstrapKoaApp} = require('./src/util');
 
 const agendaAppUrl = 'http://localhost:4041';
@@ -9,29 +10,6 @@ const {app: testApp, router: testAppRouter} = bootstrapKoaApp();
 const getTestAppUrl = path => path ? `${testAppUrl}${path}` : testAppUrl;
 
 const agendaAppRequest = request(agendaAppUrl);
-
-testAppRouter.post('/foo', async (ctx, next) => {
-  console.log('foo invoked!');
-  ctx.body = 'foo success';
-  ctx.status = 200;
-  await next();
-});
-
-testAppRouter.post('/foo/:fooParam', async (ctx, next) => {
-  console.log('foo with params invoked!');
-  console.log(ctx.params);
-  console.log(ctx.request.body);
-  ctx.body = 'foo with params success';
-  ctx.status = 200;
-  await next();
-});
-
-testAppRouter.post('/foo/cb', async (ctx, next) => {
-  console.log('foo callback invoked!');
-  ctx.body = 'foo callback success';
-  ctx.status = 200;
-  await next();
-});
 
 const bootstrapApp = async () => {
   const {app, jobsReady} = require('./src');
@@ -75,6 +53,60 @@ test.serial('PUT /api/job succeeds when the job exists', async t => {
     .send({url: getTestAppUrl('/foo')});
 
   t.is(res.status, 200);
+});
+
+const testAppCalled = async (defineEndpoint, thisManyTimes = 1) => {
+  const calls = range(thisManyTimes).map(() => new Promise(resolve => defineEndpoint(resolve)));
+  return Promise.all(calls);
+};
+
+const defineEndpointFoo = done => testAppRouter.post('/foo', async (ctx, next) => {
+  console.log('foo invoked!');
+  ctx.body = 'foo success';
+  ctx.status = 200;
+  done();
+  await next();
+});
+/* TODO
+const defineEndpointFooParam = done => testAppRouter.post('/foo/:fooParam', async (ctx, next) => {
+  console.log('foo with params invoked!');
+  console.log(ctx.params);
+  console.log(ctx.request.body);
+  ctx.body = 'foo with params success';
+  ctx.status = 200;
+  done();
+  await next();
+});
+
+const defineEndpointFooCallback = done => testAppRouter.post('/foo/cb', async (ctx, next) => {
+  console.log('foo callback invoked!');
+  ctx.body = 'foo callback success';
+  ctx.status = 200;
+  done();
+  await next();
+});
+*/
+
+test.serial('POST /api/job/now with existing foo definition invokes the foo endpoint', async t => {
+  const fooInvoked = testAppCalled(defineEndpointFoo);
+  const res = await agendaAppRequest
+    .post('/api/job/now')
+    .send({name: 'foo'});
+
+  t.is(res.text, 'job scheduled for now');
+
+  return fooInvoked;
+});
+
+test.serial('POST /api/job/every with existing foo definition invokes the foo endpoint', async t => {
+  const fooInvoked = testAppCalled(defineEndpointFoo, 3);
+  const res = await agendaAppRequest
+    .post('/api/job/every')
+    .send({name: 'foo', interval: '10 miliseconds'});
+
+  t.is(res.text, 'job scheduled for repetition');
+
+  return fooInvoked;
 });
 
 test.serial('DELETE /api/job succeeds when a job is defined', async t => {

--- a/test.js
+++ b/test.js
@@ -57,13 +57,9 @@ test.serial('PUT /api/job succeeds when the job exists', async t => {
 let counter;
 
 testAppRouter.post('/foo', async (ctx, next) => {
-  console.log('foo invoked!');
-  console.log(new Date());
   ctx.body = 'foo success';
   ctx.status = 200;
-  if (counter) {
-    counter.ready.then(() => counter.count());
-  }
+  console.log(`foo invoked! ${await counter.count()} of ${counter.countTimes} times`);
   await next();
 });
 /* TODO
@@ -94,7 +90,7 @@ test.serial('POST /api/job/now with existing foo definition invokes the foo endp
 
   t.is(res.text, 'job scheduled for now');
 
-  return counter.finished;
+  await counter.finished;
 });
 
 test.serial('POST /api/job/every with existing foo definition invokes the foo endpoint', async t => {

--- a/test.js
+++ b/test.js
@@ -56,7 +56,23 @@ test.serial('POST /api/job fails without content', async t => {
 test.serial('POST /api/job succeeds when a job is specified', async t => {
   const res = await agendaAppRequest
     .post('/api/job')
-    .send({name: 'foo', url: getTestAppUrl('/foo')});
+    .send({name: 'foo', url: getTestAppUrl('/fooWrong')});
+
+  t.is(res.status, 200);
+});
+
+test.serial('PUT /api/job fails when the job does not exists', async t => {
+  const res = await agendaAppRequest
+    .put('/api/job/fooWrong')
+    .send({url: getTestAppUrl('/foo')});
+
+  t.is(res.status, 400);
+});
+
+test.serial('PUT /api/job succeeds when the job exists', async t => {
+  const res = await agendaAppRequest
+    .put('/api/job/foo')
+    .send({url: getTestAppUrl('/foo')});
 
   t.is(res.status, 200);
 });

--- a/test.js
+++ b/test.js
@@ -1,7 +1,7 @@
 const {promisify} = require('util');
 const test = require('ava');
 const request = require('supertest');
-const {bootstrapKoaApp, AsyncCounter} = require('./src/util');
+const {bootstrapKoaApp, oncePerKey, AsyncCounter} = require('./src/util');
 
 const agendaAppUrl = 'http://localhost:4041';
 const testAppUrl = 'http://localhost:4042';
@@ -54,36 +54,44 @@ test.serial('PUT /api/job succeeds when the job exists', async t => {
   t.is(res.status, 200);
 });
 
-let counter;
+const fooProps = {};
 
-testAppRouter.post('/foo', async (ctx, next) => {
-  ctx.body = 'foo success';
-  ctx.status = 200;
-  console.log(`foo invoked! ${await counter.count()} of ${counter.countTimes} times`);
-  await next();
-});
+const defineFooEndpoint = (route, message, countTimes = 1, statusCode = 200) => {
+  const counter = new AsyncCounter(countTimes);
+  fooProps.counter = counter;
+  fooProps.message = message;
+  fooProps.statusCode = statusCode;
+
+  const define = oncePerKey(route, () => testAppRouter.post(route, async (ctx, next) => {
+    ctx.body = fooProps.message;
+    ctx.status = fooProps.statusCode;
+    console.log(`${fooProps.message}! ${await fooProps.counter.count()} of ${fooProps.counter.countTimes}`);
+    await next();
+  }));
+  define();
+  return counter;
+};
+
 /* TODO
-const defineEndpointFooParam = done => testAppRouter.post('/foo/:fooParam', async (ctx, next) => {
+testAppRouter.post('/foo/:fooParam', async (ctx, next) => {
   console.log('foo with params invoked!');
   console.log(ctx.params);
   console.log(ctx.request.body);
   ctx.body = 'foo with params success';
   ctx.status = 200;
-  done();
   await next();
 });
 
-const defineEndpointFooCallback = done => testAppRouter.post('/foo/cb', async (ctx, next) => {
+testAppRouter.post('/foo/cb', async (ctx, next) => {
   console.log('foo callback invoked!');
   ctx.body = 'foo callback success';
   ctx.status = 200;
-  done();
   await next();
 });
 */
 
 test.serial('POST /api/job/now with existing foo definition invokes the foo endpoint', async t => {
-  counter = new AsyncCounter(1);
+  const counter = defineFooEndpoint('/foo', 'foo now invoked');
   const res = await agendaAppRequest
     .post('/api/job/now')
     .send({name: 'foo'});
@@ -94,12 +102,24 @@ test.serial('POST /api/job/now with existing foo definition invokes the foo endp
 });
 
 test.serial('POST /api/job/every with existing foo definition invokes the foo endpoint', async t => {
-  counter = new AsyncCounter(3);
+  const counter = defineFooEndpoint('/foo', 'foo every invoked', 3);
   const res = await agendaAppRequest
     .post('/api/job/every')
     .send({name: 'foo', interval: '2 seconds'});
 
   t.is(res.text, 'job scheduled for repetition');
+
+  await counter.finished;
+});
+
+test.serial('POST /api/job/once with existing foo definition invokes the foo endpoint', async t => {
+  const counter = defineFooEndpoint('/foo', 'foo once invoked');
+  const res = await agendaAppRequest
+    .post('/api/job/once')
+    .send({name: 'foo', interval: new Date().getTime() + 10000});
+    // .send({name: 'foo', interval: 'in 10 seconds'});
+
+  t.is(res.text, 'job scheduled for once');
 
   await counter.finished;
 });

